ambry-tools/src/main/java/com/github/ambry/tools/perf/serverperf/ServerPerfNetworkQueue.java

```diff
@@ -0,0 +1,256 @@
+
+      Iterator<Map.Entry<Integer, Long>> pendingRequestIterator =
+          pendingCorrelationIdToStartTimeMs.entrySet().iterator();
+
+      while (pendingRequestIterator.hasNext()) {
+        Map.Entry<Integer, Long> entry = pendingRequestIterator.next();
+        int correlationId = entry.getKey();
+        long startTimeMs = entry.getValue();
+
+        long currentTimeInMs = time.milliseconds();
+
+        if (startTimeMs + operationsTimeOutMs < currentTimeInMs) {
+
+          responseInfos.add(new ResponseInfo(pendingCorrelationIdToRequestInfo.get(correlationId),
+              NetworkClientErrorCode.TimeoutError, null));
+
+          responseInfos.addAll(networkClients.get(pendingCorrelationIdToNetworkClientIdx.get(correlationId))
+              .sendAndPoll(Collections.emptyList(), new HashSet<>(Collections.singletonList(correlationId)),
+                  pollTimeout));
+          pendingRequestIterator.remove();
+        } else {
+          break;
+        }
+      }
+
+      networkClients.forEach(networkClient -> {
+        responseInfos.addAll(networkClient.sendAndPoll(Collections.emptyList(), new HashSet<>(), pollTimeout));
+      });
+
+      if (!requestInfos.isEmpty()) {
+        RequestInfo requestInfo = requestInfos.peek();
+
+        pendingCorrelationIdToStartTimeMs.put(requestInfo.getRequest().getCorrelationId(), time.milliseconds());
+        pendingCorrelationIdToNetworkClientIdx.put(requestInfo.getRequest().getCorrelationId(), clientIndex);
+        pendingCorrelationIdToRequestInfo.put(requestInfo.getRequest().getCorrelationId(), requestInfo);
+        requestInfos.remove();
+
+        List<ResponseInfo> responses = networkClients.get(clientIndex)
+            .sendAndPoll(Collections.singletonList(requestInfo), new HashSet<>(), pollTimeout);
+        clientIndex++;
+        clientIndex = clientIndex % networkClients.size();
+        responseInfos.addAll(responses);
+      }
+
+      this.responseInfos.addAll(responseInfos);
+
+      responseInfos.forEach(responseInfo -> {
+        pendingCorrelationIdToStartTimeMs.remove(responseInfo.getRequestInfo().getRequest().getCorrelationId());
+        pendingCorrelationIdToNetworkClientIdx.remove(responseInfo.getRequestInfo().getRequest().getCorrelationId());
+        pendingCorrelationIdToRequestInfo.remove(responseInfo.getRequestInfo().getRequest().getCorrelationId());
+      });
+    }
+
+    try {
+      maxParallelRequest.release(1);
+      maxParallelRequest.acquire(maxParallelism + 1);
+    } catch (InterruptedException e) {
+      logger.warn("Caught exception while waiting for executor service to finish", e);
+    } finally {
+      executorService.shutdownNow();
+      shutDownLatch.countDown();
+    }
+  }
+
+  void shutDown() throws Exception {
+    isShutDown = true;
+    shutDownLatch.await();
+  }
+
+  @Override
+  public void close() throws IOException {
+    networkClients.forEach(Http2NetworkClient::close);
+  }
```

Comment:
we should merge this two methods, right, we need to close the network client when the PerformanceQueue is shutting down.

Reply:
Merged it.